### DISCUSSION TITLE
Ambiguity between aliases in context passed to DataLoaders

### DISCUSSION THREAD

**Discussion by nick-stephen** [graphql-java/graphql-java#2462] [Q&A]
I'm using aliases in order to specify two different sets of parameters to a navigation in the same query:

```
shows {
  title
  best: reviews(top: 3) {
     score
  }
  worst: reviews(bottom: 3) {
     score
  }
}
```

The implementation of 'shows' query and the 'reviews' field have independent `DataFetchers` and `DataLoaders`.

The `ReviewsDataFetcher` is being invoked twice for each show (once for each alias) each time with the 'Show' object in the `DataFetchingEnvironment.getSource()`

My code invokes the `ReviewsDataLoader` with the source Show's ID as the key and with the `DataFetchingEnvironment.getArguments()` as the `keyContext`.

In the `ReviewDataLoader`, which implements `MappedBatchLoaderWithContext`, the problem is that the `environment.getKeyContexts()` map has overwritten the different arguments from the two different aliases, since they both use the same show key. So we only get one entry per show. 

The query results end up with the same values for both the 'best' and 'worst' aliases.

After some investigation it seems incorrect that the `FieldCoordinates` for the two aliases are the same as each other - this appears to cause re-use of the same `DataFetcher` and `DataLoader` across the two aliases despite the sub-tree of each alias potentially being different and having different parameters. It would seem that a fix would be to add the 'alias' field to the FieldCoordinates, so that the two aliases are distinguished from each other.

I can work around this by using a compound 'key' in the DataFetcher/Loader which also includes the alias name, and splitting the input in the DataLoader by alias to perform multiple separate queries in parallel threads, and join the results, but this adds complexity and seems to be a workaround to a bug - typically I'd expect a DataLoaders to be invoked once for each coordinate in the query tree, and there are two coordinates here.

This seems to be related but an opposite use-case to issue: https://github.com/graphql-java/graphql-java/issues/1537

In my implementation, I'm using the DGS framework and Apollo Federation, with the 'reviews' field being federated into the 'shows' object, not sure if that makes any difference. 

Adding callout to @jmccaull who appears to have been involved in responding to the above issue... Thx!

**Comment by bbakerman** [comment]
aliases are a query side concept.  The code making the query can use aliases or not use them.

They do NOT form part of the schema contract and MUST not be accounted for in the back end.  FieldCoorinates etc... are about data fetching code to types and fields.  Aliases will never factor into them.

Since you want to have a best review and worse review - you should create ` bestReviews` field and a `worseReviews` field - not relying on query side aliases to be your differentiator

As for DataLoader - it requires that the key you can be unique - so even side from the above, you would need to create a compound key that is `$showId+best` and `$showId+worst` so that calls are unique.

The data loader does not use key + key context to make a unique entry - its only ever just the key.  So your keys must be unique if you want different data semantics

None of this is a bug - and I suggest your DONT aliases to differentiate field meaning - use new fields.

```
type Show {
   id  : ID
   bestReviews( top : Int) [Review]
   worseReviews( bottom : Int) [Review]   
}
```

**Reply by nick-stephen** [reply]
Thanks for the prompt reply!

I understand what you're saying, and will continue to use a compound key to address this issue, but I'm not sure I agree that we should always address the above kind of use-case with separate fields - this was just an artificial example - any field where filtering or other parameters are passed in is a candidate for the client wanting to request results for two or more possible sets of filters / arguments, and all the possibilities can't be planned for in the graphql schema - unless you're suggesting that we move to a model where each field that takes arguments instead takes argument arrays, allowing multiple choices to be selected by the client, for the same field. This renders aliases completely unnecessary. An example might be that the UI wants to display several different metrics fetched with different parameters - the schema developer does not know ahead of time know how many different metrics the UI may ask for or which time periods, hence these are specified in query parameters, and aliasing allows the front-end to optimize its requests by fetching multiple metrics for the same resource, with different parameters, using aliases. It's not as binary as best/worst, which I agree could have been planned for in the schema. The client may specify 0, 1, 2 or 3 or more different groups of metrics with different parameters for a given parent, all in the same request, via use of aliases for each grouping. Indeed, this is what aliases are there for as I'd understood things - otherwise what use are they at all?

In any case, you've answered the question, which is that this is by design, and that it's the responsibility of the backend to generate composite keys, separate out the different keys in the loader, run appropriate queries, and aggregate the results, dealing with any parallelism on its own, rather than the infrastructure isolating different parts of the graph query into different dataloaders based not only the position in the schema but also on use of aliases with different parameters.

**Comment by bbakerman** [comment]
You can of course build a system where the DataFetchers recognise aliases.
You dont have to use different fields as I suggested.

However now you have the case where the input of the consumer dictates the
behavior of the server.  This might be ok in your case because you have
deep control over both the query maker and the query taker.

But if you are trying to write  a general purpose API, then it's not a
great recipe for success

**Reply by nick-stephen** [reply]
I'd rather not need the code understand the aliases, hence the thread. But there are multiple different queries coming from the client with different arguments, via use of aliases, and when the client can provide arguments in a graphql query, this obviously dictates some behaviour to the server. 

I think this conversation is helping me understand this a little more deeply. Rather than including the alias in the key used by the DataLoaders, it would be more correct to include the entity ID and the arguments map in the key, and not use the key context for the arguments. 

By putting both the entity Id and the arguments in the key passed in to the DataLoader, this would enable the caching mechanism to work across queries in different aliases any time that arguments are shared (which I'd expect not to happen in our use-case, but it might) - but still allowing the backend to run different queries for the same entity/ies if there are more than one alias with different arguments. 

The onus would still remain on the DataLoader to group and run multiple backend queries to DB or wherever in order to fulfill the results from each of the aliases into one response, and to parallelise this as appropriate, but our code is already doing that.

Thx for your advice, much appreciated.
